File: static/solutions/java/src/main/java/com/epi/TeamReachability.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

package com.epi;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Map;
import java.util.Set;

public class TeamReachability {
  // @include
  public static class MatchResult {
    public String winningTeam;
    public String losingTeam;

    public MatchResult(String winningTeam, String losingTeam) {
      this.winningTeam = winningTeam;
      this.losingTeam = losingTeam;
    }
  }

  public static boolean canTeamABeatTeamB(List<MatchResult> matches,
                                          String teamA, String teamB) {
    Set<String> visited = new HashSet<>();
    return isReachableDFS(buildGraph(matches), teamA, teamB, visited);
  }

  private static Map<String, Set<String>> buildGraph(
      List<MatchResult> matches) {
    Map<String, Set<String>> graph = new HashMap<>();
    for (MatchResult match : matches) {
      Set<String> edges = graph.get(match.winningTeam);
      if (edges == null) {
        edges = new HashSet<>();
        graph.put(match.winningTeam, edges);
      }
      edges.add(match.losingTeam);
    }
    return graph;
  }

  private static boolean isReachableDFS(Map<String, Set<String>> graph,
                                        String curr, String dest,
                                        Set<String> visited) {
    if (curr.equals(dest)) {
      return true;
    } else if (visited.contains(curr) || graph.get(curr) == null) {
      return false;
    }
    visited.add(curr);
    for (String team : graph.get(curr)) {
      if (isReachableDFS(graph, team, dest, visited)) {
        return true;
      }
    }
    return false;
  }
  // @exclude

  private static boolean isReachableBFS(Map<String, Set<String>> graph,
                                        String curr, String dest,
                                        Set<String> visited) {
    Queue<String> frontier = new LinkedList<>();
    visited.add(curr);
    frontier.add(curr);
    while (!frontier.isEmpty()) {
      String justReached = frontier.poll();
      if (justReached.equals(dest)) {
        return true;
      } else if (graph.get(justReached) != null) {
        for (String team : graph.get(justReached)) {
          if (!visited.contains(team)) {
            visited.add(team);
            frontier.add(team);
          }
        }
      }
    }
    return false;
  }

  private static void check(List<MatchResult> matches, String teamA,
                            String teamB) {
    Set<String> visited = new HashSet<>();
    assert(canTeamABeatTeamB(matches, teamA, teamB)
           == isReachableBFS(buildGraph(matches), teamA, teamB, visited));
  }

  public static void main(String[] args) {
    List<MatchResult> matches = Arrays.asList(
        new MatchResult("Texas", "Cal"), new MatchResult("Cal", "Stanford"),
        new MatchResult("Stanford", "Texas"),
        new MatchResult("Stanford", "MIT"), new MatchResult("Stanford", "CIT"),
        new MatchResult("UCLA", "USC"));
    check(matches, "Texas", "MIT");
    check(matches, "Cal", "UCLA");
  }
}

File: static/python/team_reachability.py
import collections

# @include
MatchResult = collections.namedtuple('MatchResult',
                                     ('winning_team', 'losing_team'))


def can_team_a_beat_team_b(matches, team_a, team_b):
    def build_graph():
        graph = collections.defaultdict(set)
        for match in matches:
            graph[match.winning_team].add(match.losing_team)
        return graph

    def is_reachable_dfs(graph, curr, dest, visited=set()):
        if curr == dest:
            return True
        elif curr in visited or curr not in graph:
            return False
        visited.add(curr)
        return any(is_reachable_dfs(graph, team, dest) for team in graph[curr])

    return is_reachable_dfs(build_graph(), team_a, team_b)
# @exclude


def check(matches, team_a, team_b):
    def build_graph():
        graph = collections.defaultdict(set)
        for match in matches:
            graph[match.winning_team].add(match.losing_team)
        return graph

    def is_reachable_bfs(graph, curr, dest, visited=set()):
        frontier = collections.deque([curr])
        visited.add(curr)
        while frontier:
            just_reached = frontier.popleft()
            if just_reached == dest:
                return True
            elif just_reached in graph:
                for team in graph[just_reached]:
                    if team not in visited:
                        visited.add(team)
                        frontier.append(team)
        return False

    assert can_team_a_beat_team_b(matches, team_a, team_b) == is_reachable_bfs(
        build_graph(), team_a, team_b)


def main():
    matches = [
        MatchResult('Texas', 'Cal'), MatchResult('Cal', 'Stanford'),
        MatchResult('Stanford', 'Texas'), MatchResult('Stanford', 'MIT'),
        MatchResult('Stanford', 'CIT'), MatchResult('UCLA', 'USC')
    ]
    check(matches, 'Texas', 'MIT')
    check(matches, 'Cal', 'UCLA')


if __name__ == '__main__':
    main()
